jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/message/MessageInputStream.java

```diff
@@ -130,47 +145,53 @@ public void close()
                     return;
 
                 case SUSPENDED:
-                    resume = suspendToken;
-                    suspendToken = null;
-                    state = State.CLOSED;
-                    break;
-
                 case RESUMED:
-                    state = State.CLOSED;
+                    state = State.COMPLETE;
                     break;
+
+                default:
+                    throw new IllegalStateException();
             }
 
-            buffers.clear();
             buffers.offer(EOF);
         }
-
-        // May need to resume to discard until we reach next message.
-        if (resume != null)
-            resume.resume();
     }
 
-    @Override
-    public void mark(int readlimit)
+    public void handlerComplete()
     {
-        // Not supported.
-    }
+        // Close the InputStream.
+        close();
 
-    @Override
-    public boolean markSupported()
-    {
-        return false;
+        // May need to resume to resume and read to the next message.
+        SuspendToken resume;
+        synchronized (this)
+        {
+            resume = suspendToken;
+            suspendToken = null;
+        }
+
+        if (resume != null)
+            resume.resume();
     }
 
     @Override
-    public void messageComplete()
+    public int read() throws IOException
     {
-        if (LOG.isDebugEnabled())
-            LOG.debug("Message completed");
-        buffers.offer(EOF);
+        byte[] bytes = new byte[1];
+        while (true)
+        {
+            int read = read(bytes, 0, 1);
+            if (read < 0)
+                return -1;
+            if (read == 0)
+                continue;
+
+            return bytes[0] & 0xFF;
```

Comment:
We loop infinitely on COMPLETE?
If so, I'm not liking this solution.

Reply:
If the message is complete, that should offer the last frame with FIN=true and then an EOF in the buffer queue.
Why do we have a new state?

Reply:
The state `COMPLETE` means that we have all the content of the message and have appended an EOF buffer in the queue after that. We need this state so we don't resume reading new frames after the `fin==true` and to enforce that no new frames are appended after the `fin==true`.

We will not loop infinitely on COMPLETE, we will read `-1` when we run out of content to read.